### PR TITLE
fix(tree-entity): closure junction table primary key definition should match parent table

### LINKED ISSUE #9600: [BUG]: if set closure-table primary key unsigned, it occurs FK error
### Issue description

I want to use closure-table and if I set primary key ( unsigned ) , error comes out when migrate shcema

### Expected Behavior

```
@Entity()
@Tree('closure-table', {
  closureTableName: 'category',
  ancestorColumnName: (column) => 'ancestor_' + column.propertyName,
  descendantColumnName: (column) => 'descendant_' + column.propertyName,
})
export class CategoryEntity {
  @PrimaryGeneratedColumn('increment', { type: 'bigint', unsigned: true })
  id: string
}
```
I defined CategoryEntity like this.
And I expected closure-table is created successfully when create table by typeorm migration

### Actual Behavior

when migrate schema, error occures below
```
error: Error: Referencing column 'ancestor_id' and referenced column 'id' in foreign key constraint 'FK_ab837edaafe1cd64a7efb216db1' are incompatible.
migration-file         |     at Packet.asError (/usr/src/app/node_modules/mysql2/lib/packets/packet.js:728:17)
migration-file         |     at Query.execute (/usr/src/app/node_modules/mysql2/lib/commands/command.js:29:26)
migration-file         |     at PoolConnection.handlePacket (/usr/src/app/node_modules/mysql2/lib/connection.js:456:32)
migration-file         |     at PacketParser.onPacket (/usr/src/app/node_modules/mysql2/lib/connection.js:85:12)
migration-file         |     at PacketParser.executeStart (/usr/src/app/node_modules/mysql2/lib/packet_parser.js:75:16)
migration-file         |     at Socket.<anonymous> (/usr/src/app/node_modules/mysql2/lib/connection.js:92:25)
migration-file         |     at Socket.emit (events.js:400:28)
migration-file         |     at Socket.emit (domain.js:475:12)
migration-file         |     at addChunk (internal/streams/readable.js:293:12)
migration-file         |     at readableAddChunk (internal/streams/readable.js:267:9) {
migration-file         |   code: undefined,
migration-file         |   errno: 3780,
migration-file         |   sqlState: 'HY000',
migration-file         |   sqlMessage: "Referencing column 'ancestor_id' and referenced column 'id' in foreign key constraint 'FK_ab837edaafe1cd64a7efb216db1' are incompatible.",
migration-file         |   sql: 'ALTER TABLE `category_closure` ADD CONSTRAINT `FK_ab837edaafe1cd64a7efb216db1` FOREIGN KEY (`ancestor_id`) REFERENCES `category`(`id`) ON DELETE CASCADE ON UPDATE NO ACTION'
```

### Steps to reproduce

set primary column option unsigned
@PrimaryGeneratedColumn('increment', { type: 'bigint', unsigned: true })

and do migraion. can reproduce

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Node.js version     | 16.4.0  |
| TypeORM version     | 0.3.11 |

### Additional Context

how can I resolve this issue?

can't set any options at closure-table except tableName, ancestorColumnName, descendantColumnName

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, and I know how to start.

### PR DESCRIPTION
**Pull request by gongAll**
### Description of change

Bugfix of behaviour described in issue #9600

### Pull-Request Checklist

- [ x ] Code is up-to-date with the `master` branch
- [ x ] `npm run format` to apply prettier formatting
- [ x ] `npm run test` passes with this change
- [ x ] This pull request links relevant issues as `Fixes #0000`
- [ N/A ] There are new or updated unit tests validating the change
- [ N/A ] Documentation has been updated to reflect this change
- [ x ] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

## Summary by CodeRabbit

- **New Features**
  - Added new entity types supporting closure-table trees with customizable column attributes, including unsigned integers, decimals with formatting options, and varchars with charset/collation settings.
- **Tests**
  - Introduced functional tests to verify that closure table columns correctly inherit detailed properties such as unsigned, zerofill, width, precision, scale, charset, and collation from their referenced primary columns.

### CODE REVIEW

**Review by gioboa — COMMENTED** [review]
Thanks @gongAll for your help.
Can you create a test to cover this fix please?

**Comment by gongAll** [comment]
@gioboa I was trying to, but am having some difficulty setting up the database locally to run my test.

Getting 

`Error: Connection lost: The server closed the connection.`

While trying to run test with mysql database. Executed your docker compose for the mysql database (only changed port to 3307) and setted the ormconfig as described in docs, as follows:

```
[
  {
    "skip": false,
    "name": "mysql",
    "type": "mysql",
    "host": "localhost",
    "port": 3307,
    "username": "root",
    "password": "admin",
    "database": "test",
    "logging": true
  }
]
```

**Review by gioboa — COMMENTED** [review]
I see. You can commit your test and the pipeline will run it for you.

**Comment by gongAll** [comment]
@gioboa ok managed to resolve problems on my side and implemented tests.

**Comment by gongAll** [comment]
@gioboa i see some tests have failed, but I can't quite get a grasp of what it is that went wrong. Could you give me any feedback regarding this? Cheers.

**Review by gioboa — COMMENTED** [review]
@alumni Is there a way to fix these failing tests?

```
Error: Unable to download artifact(s): Artifact not found for name: build
        Please ensure that your artifact is not expired and the artifact was uploaded using a compatible version of toolkit/upload-artifact.
```

**Comment by gongAll** [comment]
@gioboa any news on this? From that error I don't understand what I can do on my side to resolve the failing tests, perhaps someone more experienced in working with the library can chime in with some hints?

**Review by gioboa — COMMENTED** [review]
I aligned this PR with master. It's working fine now 👏

**Comment by alumni** [comment]
Should there be any other column options we should consider?

**Review by naorpeled — APPROVED** [review]

File: test/github-issues/9600/issue-9600.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    createTestingConnections,
+    reloadTestingDatabases,
+} from "../../utils/test-utils"
+import { DataSource } from "../../../src"
+import { FooEntity } from "./entity/Foo"
+import { expect } from "chai"
+
+// if bug persists, will throw error when creating testing connection, won't get to the test
+
+describe("github issues > #9600 if set closure-table primary key unsigned, it occurs FK error", () => {
```
Comment (sgarner) [inline test/github-issues/9600/issue-9600.ts:13]: We're moving towards placing tests in the [functional tests directory](https://github.com/typeorm/typeorm/tree/master/test/functional) since it's easier to see what is covered and not duplicate tests.

We already have `test/functional/tree-tables/closure-table/closure-table.ts`where these tests could go.

**Review by alumni — APPROVED** [review]

**Review by gioboa — APPROVED** [review]
Thanks @gongAll 👏

**Comment by gongAll** [comment]
@alumni 
> Should there be any other column options we should consider?

I will leave those considerations up to you guys, for my use case the unsigned is enough although there might exist some weird edge cases which probably will arise from lack of good practices.

@sgarner 
> We're moving towards placing tests in the [functional tests directory](https://github.com/typeorm/typeorm/tree/master/test/functional) since it's easier to see what is covered and not duplicate tests.

Shouldn't be too hard, just wanna make sure what exactly is best:
- remove from where it is and add to functional tests
- maintain as is and add to functional tests

I can see the reason why both approaches are good, so I leave that up to you.

**Comment by sgarner** [comment]
@gongAll Since we're here, let's copy all of the relevant column properties to the closure table key:

- [x] length
- [x] unsigned
- [ ] width
- [ ] precision
- [ ] scale
- [ ] zerofill
- [ ] charset
- [ ] collation

At least for MySQL, both collation and charset are significant for character column types:

[image](https://dev.mysql.com/doc/refman/8.4/en/create-table-foreign-keys.html)

**Comment by gongAll** [comment]
@gioboa @alumni @sgarner just pinging to get an update regarding this PR.

File: test/functional/tree-tables/closure-table/closure-table.ts [hunk 2]
```diff
@@ -728,4 +732,115 @@ describe("tree tables > closure-table", () => {
                 })
             }),
         ))
+
+    it("foo1 should create closure columns unsigned", () =>
+        Promise.all(
+            connections.map(async (dataSource) => {
```
Comment (sgarner) [inline test/functional/tree-tables/closure-table/closure-table.ts]: These entities use mysql-specific column options, causing the tests to fail on other databases.

Probably the easiest answer is to limit these tests to only run on mysql:

```suggestion
    it("foo1 should create closure columns unsigned", () =>
        Promise.all(
            connections.map(async (dataSource) => {
                if (dataSource.driver.options.type !== "mysql") {
                    return
                }
```

File: test/functional/tree-tables/closure-table/closure-table.ts [hunk 3]
```diff
@@ -728,4 +732,115 @@ describe("tree tables > closure-table", () => {
...
+                expect(descendantCol).to.exist
+
+                expect(ancestorCol.unsigned).to.be.true
+                expect(descendantCol.unsigned).to.be.true
+            }),
+        ))
+
+    it("foo2 should create closure columns with specified zerofill, width, precision and scale", () =>
+        Promise.all(
+            connections.map(async (dataSource) => {
```
Comment (sgarner) [inline test/functional/tree-tables/closure-table/closure-table.ts]:
```suggestion
    it("foo2 should create closure columns with specified zerofill, width, precision and scale", () =>
        Promise.all(
            connections.map(async (dataSource) => {
                if (dataSource.driver.options.type !== "mysql") {
                    return
                }
```

File: test/functional/tree-tables/closure-table/closure-table.ts [hunk 4]
```diff
@@ -728,4 +732,115 @@ describe("tree tables > closure-table", () => {
...
+                expect(descendantCol.precision).to.be.eq(9)
+
+                expect(ancestorCol.scale).to.be.eq(3)
+                expect(descendantCol.scale).to.be.eq(3)
+            }),
+        ))
+
+    it("foo3 should create closure columns with specified length, charset and collation", () =>
+        Promise.all(
+            connections.map(async (dataSource) => {
```
Comment (sgarner) [inline test/functional/tree-tables/closure-table/closure-table.ts]:
```suggestion
    it("foo3 should create closure columns with specified length, charset and collation", () =>
        Promise.all(
            connections.map(async (dataSource) => {
                if (dataSource.driver.options.type !== "mysql") {
                    return
                }
```

**Review by sgarner — CHANGES_REQUESTED** [review]

**Comment by gongAll** [comment]
@sgarner i appreciate your feedback, just implemented the requested changes.

**Comment by sgarner** [comment]
Sorry, the tests are still failing, the last fix wasn't sufficient. 😞 

The problem is the `before`/`beforeEach` hooks are still trying to create the problematic entities for all connections.

We may have to split these new tests into a separate `describe() {...}` block with their own hooks. Then you can use the `enabledDrivers` option in `createTestingConnections` to limit these tests to mysql, including the database setup phase. The `FooX` entities should only be listed for these tests.

**Comment by gongAll** [comment]
@sgarner seems to have passed the tests now, let me know if any other action is necessary. :smile:

**Review by alumni — APPROVED** [review]

**Review by gioboa — APPROVED** [review]

**Review by sgarner — APPROVED** [review]
Thanks for your contribution @gongAll 💜

**Comment by td-worzel** [comment]
@sgarner @gongAll 

Hi, thanks for addressing this issue. Unfortunately, this change is breaking for existing applications that rely on the closure table being created in a database different from the parent entity.

We upgraded TypeORM and discovered that our existing implementation stopped working because the closure table database is now derived from the parent entity instead of using the globally configured database. While it is technically possible to migrate the affected tables to the parent entity's database, this is risky for us because the entity is widely used across our platform and such a migration would have a significant impact.

I previously proposed an alternative approach in https://github.com/typeorm/typeorm/issues/11121, which, as far as I can see, is also the approach used in the 1.0 version. We are not currently in a position to migrate to 1.0, though.

Would it be possible to consider a follow-up change that provides an explicit option for selecting the database/schema in which the closure table should be created? This would preserve the current behaviour by default while allowing applications that need the closure table in a different database to configure it explicitly.

Archived issue: https://github.com/typeorm/typeorm/issues/11121
